gr-fec/lib/async_decoder_impl.cc

```diff
@@ -112,88 +107,56 @@ void async_decoder_impl::decode_unpacked(pmt::pmt_t msg)
 
     size_t o0(0);
     const float* f32in = pmt::f32vector_elements(bits, o0);
-    pmt::pmt_t outvec(pmt::make_u8vector(nbits_out, 0x00));
+    const size_t nitems_out = d_packed ? nbits_out / 8 : nbits_out;
+    pmt::pmt_t outvec(pmt::make_u8vector(nitems_out, 0x00));
     uint8_t* u8out = pmt::u8vector_writable_elements(outvec, o0);
+    uint8_t* decoder_out = d_packed ? d_bits_out.data() : u8out;
+    const float shift = d_decoder->get_shift();
 
-    if (strncmp(d_decoder->get_input_conversion(), "uchar", 5) == 0) {
+    if (std::string_view(d_decoder->get_input_conversion()) == "uchar") {
+        // TODO: replace all this by volk_32f_s32f_x2_convert_8u when it gets
+        // into a stable volk release
         volk_32f_s32f_multiply_32f(d_tmp_f32.data(), f32in, 48.0f, nbits_in);
-    } else {
-        memcpy(d_tmp_f32.data(), f32in, nbits_in * sizeof(float));
-    }
-
-    if (d_decoder->get_shift() != 0) {
-        for (size_t n = 0; n < nbits_in; n++)
-            d_tmp_f32[n] += d_decoder->get_shift();
-    }
-
-    if (strncmp(d_decoder->get_input_conversion(), "uchar", 5) == 0) {
-        // volk_32f_s32f_convert_8i(d_tmp_u8, d_tmp_f32, 1, nbits_in);
-        for (size_t n = 0; n < nbits_in; n++)
-            d_tmp_u8[n] = static_cast<uint8_t>(d_tmp_f32[n]);
+        if (shift != 0.0) {
+            for (size_t n = 0; n < nbits_in; n++) {
+                d_tmp_f32[n] += shift;
+            }
+        }
+        for (size_t n = 0; n < nbits_in; n++) {
+            d_tmp_u8[n] = std::clamp<float>(d_tmp_f32[n], 0.0f, UINT8_MAX);
+        }
 
-        d_decoder->generic_work((void*)d_tmp_u8.data(), (void*)u8out);
+        for (size_t i = 0; i < nblocks; i++) {
+            d_decoder->generic_work(
+                (void*)&d_tmp_u8[i * d_decoder->get_input_size()],
+                (void*)&decoder_out[i * d_decoder->get_output_size()]);
+        }
     } else {
+        if (shift != 0.0) {
+            volk_32f_s32f_add_32f(d_tmp_f32.data(), f32in, shift, nbits_in);
+        } else {
+            memcpy(d_tmp_f32.data(), f32in, nbits_in * sizeof(float));
+        }
+
         for (size_t i = 0; i < nblocks; i++) {
-            d_decoder->generic_work((void*)&d_tmp_f32[i * d_decoder->get_input_size()],
-                                    (void*)&u8out[i * d_decoder->get_output_size()]);
+            d_decoder->generic_work(
+                (void*)&d_tmp_f32[i * d_decoder->get_input_size()],
+                (void*)&decoder_out[i * d_decoder->get_output_size()]);
         }
     }
 
+    // Note: when nblocks > 1, this only gets the iterations for the last
+    // decoder block.
     static const pmt::pmt_t iterations_key = pmt::mp("iterations");
```

Comment:
Could you make this a class constant? Or a namespace constant? Other code in GR handles this kind of constants this way. It'd be useful for people who want to figure out how to deal with this info.

Reply:
I've moved this to the `.h` inside the `namespace fec`. I tried to move it into the `private:` for the class declaration and this doesn't work (according to [this](https://www.stroustrup.com/bs_faq2.html#in-class) it seems this will only work for integral types).

I've also changed the name to all caps (`ITERATIONS_KEY`) for consistency with some examples in gr-soapy and gr-uhd. However, I've noticed that these examples declare the keys in the `.h` outside of the namespace. If this would also be preferred here, let me know and I'll change it.

Reply:
I had these examples in mind. Intuitively, I'd expect the gr-uhd and gr-soapy keys in their namespaces. But there may be other considerations as well.